src/game_api/state.hpp

```diff
@@ -249,27 +251,42 @@ struct StateMemory
     QuestsInfo* quests;
     AITarget* ai_targets; // e.g. hired hand uid -> snake uid
     LiquidPhysics* liquid_physics;
-    std::vector<ParticleEmitterInfo*>* particle_emitters;
-    std::vector<Illumination*>* lightsources;
+    custom_vector<ParticleEmitterInfo*>* particle_emitters;
+    custom_vector<Illumination*>* lightsources;
     EntityLookup* entity_lookup;
 
     // This is a Robin Hood Table
     uint32_t uid_to_entity_mask;
     uint32_t padding13;
     RobinHoodTableEntry* uid_to_entity_data;
 
-    custom_vector<std::pair<Entity*, uint8_t>> backlayer_player_related1; // inside vector: player and destination layer?
-    uint32_t layer_transition_effect_timer;
-    /// The currently drawn layer, can't be changed
-    uint8_t camera_layer;
+    custom_vector<std::pair<Entity*, uint8_t>> entities_switching_layer; // inside vector: entity and destination layer
+                                                                         // only entities that go thru the layer (char_*, ghost, doesn't care about held items)
+                                                                         // does not care about the entity:set_layer() either
+
+    uint32_t layer_transition_timer;
+    uint8_t transition_to_layer;
     uint8_t unknown31a; // padding probably
     uint8_t unknown31b;
     uint8_t unknown31c;
     /// Holds info about owned rooms and items (shops, challenge rooms, vault etc.)
     RoomOwnersInfo room_owners;
     /// Number of frames since the game was launched
     uint32_t time_startup;
-    uint32_t special_visibility_flags;
+
+    union
+    {
+        uint32_t special_visibility_flags; // it's actually four 8bit values, last one is not even a flag, it's just layer
+
+        struct
+        {
+            uint8_t crust_visibility;
+            uint8_t compass_visibility;
+            uint8_t special_compass_visibility;
+            uint8_t camera_layer;

```

Comment:
Potentially breaking change?

Reply:
I'd say it's a fixing change. There's only one frame difference between changing transition_to_layer and camera_layer anyway, but I think it's better that the familiar camera_layer is in sync with player.layer, like it now is.